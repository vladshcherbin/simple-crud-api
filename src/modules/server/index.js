import http, { STATUS_CODES } from 'http'
import parseBody from '../body-parser'
import { ClientError, NotFoundError } from '../errors'
import pathToRegex from '../path-to-regex'
import { ValidationError } from '../validation'

export default function createServer() {
  const routes = []
  const server = http.createServer(async (request, response) => {
    response.setHeader('Content-Type', 'application/json')

    try {
      const { pathname } = new URL(request.url, 'http://127.0.0.1')
      const matchedRoute = [...routes].reverse().find(({ method, regex }) => (
        request.method === method && regex.test(pathname)
      ))

      if (matchedRoute) {
        request.params = matchedRoute.regex.exec(pathname).groups || {}
        request.body = await parseBody(request)
        response.send = (data) => {
          response.end(JSON.stringify(data))
        }

        matchedRoute.handler(request, response)
      } else {
        response.statusCode = 404
        response.end(JSON.stringify({ message: STATUS_CODES[response.statusCode] }))
      }
    } catch (error) {
      if (error instanceof ClientError) {
        response.statusCode = error.status
        response.end(JSON.stringify({ message: error.message }))
      } else if (error instanceof ValidationError) {
        response.statusCode = error.status
        response.end(JSON.stringify({ message: error.message, errors: error.errors }))
      } else if (error instanceof NotFoundError) {
        response.statusCode = 404
        response.end(JSON.stringify({ message: error.message }))
      } else {
        response.statusCode = 500
        response.end(JSON.stringify({ message: STATUS_CODES[response.statusCode] }))
      }
    }
  })

  return {
    get: (path, handler) => {
      routes.push({ method: 'GET', regex: pathToRegex(path), handler })
    },
    post: (path, handler) => {
      routes.push({ method: 'POST', regex: pathToRegex(path), handler })
    },
    put: (path, handler) => {
      routes.push({ method: 'PUT', regex: pathToRegex(path), handler })
    },
    delete: (path, handler) => {
      routes.push({ method: 'DELETE', regex: pathToRegex(path), handler })
    },
    instance() {
      return server
    },
    listen: (...args) => (
      server.listen(...args)
    )
  }
}
