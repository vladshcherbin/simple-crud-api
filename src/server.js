import createServer from './modules/server'
import checkUUID from './modules/uuid'
import * as person from './repositories/person'

const server = createServer()

server.get('/person', (_, response) => {
  response.send(person.all())
})

server.get('/person/:id', (request, response) => {
  checkUUID(request.params.id)

  response.send(person.findById(request.params.id))
})

server.post('/person', (request, response) => {
  const createdPerson = person.create(request.body)

  response.statusCode = 201
  response.send(createdPerson)
})

server.put('/person/:id', (request, response) => {
  checkUUID(request.params.id)

  response.send(person.update(request.params.id, request.body))
})

server.delete('/person/:id', (request, response) => {
  checkUUID(request.params.id)

  const removedPerson = person.remove(request.params.id)

  response.statusCode = 204
  response.send(removedPerson)
})

server.get('/missing-handler')

export default server
