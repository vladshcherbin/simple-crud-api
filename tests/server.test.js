import request from 'supertest'
import server from '../src/server'

test('Return 404 when no matching endpoint', async () => {
  const response = await request(server.instance()).get('/some-non/existing/resource')

  expect(response.status).toBe(404)
  expect(response.type).toBe('application/json')
  expect(response.body).toMatchSnapshot()
})

test('Return 500 when route handler is missing', async () => {
  const response = await request(server.instance()).get('/missing-handler')

  expect(response.status).toBe(500)
  expect(response.type).toBe('application/json')
  expect(response.body).toMatchSnapshot()
})

test('Return 400 request body is invalid', async () => {
  const response = await request(server.instance()).post('/person').send('{')

  expect(response.status).toBe(400)
  expect(response.type).toBe('application/json')
  expect(response.body).toMatchSnapshot()
})
