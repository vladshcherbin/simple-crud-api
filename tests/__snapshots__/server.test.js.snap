// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Return 400 request body is invalid 1`] = `
Object {
  "message": "Bad Request",
}
`;

exports[`Return 404 when no matching endpoint 1`] = `
Object {
  "message": "Not Found",
}
`;

exports[`Return 500 when route handler is missing 1`] = `
Object {
  "message": "Internal Server Error",
}
`;
